### PR TITLE
feat: recurring event support for change controller

### PR DESCRIPTION
**Pull request by pnwpedro**
This change implements this feature request: https://github.com/awslabs/aws-delivlib/issues/331

#### Motivation & Use Case
We use the ChangeController to enable and disable stage transitions in our pipeline based on a calendar of events. This works well when events are singular, however we'd also like to enforce change control on recurring events. For instance, we want to block stage transition into production waves after hours and on weekends.

Sample event for weekends:
```
BEGIN:VEVENT
RRULE:FREQ=WEEKLY;INTERVAL=1
DTEND:20200511T170000Z
SUMMARY:Outside of business hours (weekend)
DTSTAMP:20200504T163641Z
DTSTART:20200508T220000Z
SEQUENCE:0
END:VEVENT
```

-----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: change-control-lambda/lib/time-window.ts [hunk 1]
```diff
@@ -40,10 +42,66 @@ export function shouldBlockPipeline(icalData: string | Buffer, now = new Date(),
...
+ * Returns the previous and next events for a recurring event surrounding the
+ * provided date. If the date provided is equal to the start of an event, 
+ * the event for that date and the following event will be returend
+ * 
+ * If the provided event is not recurring, no events are returned.
+ *
+ * @param event a recurring calendar event.
+ * @param date the date for which the previous and next event should be returned.
+ */
+function getRecurringEvents(event: CalendarEvent, date: Date): CalendarEvent[] {
```
Comment (nija-at) [inline change-control-lambda/lib/time-window.ts:55]: Better naming: `flattenRecurringEvent()`

File: change-control-lambda/lib/time-window.ts [hunk 2]
```diff
@@ -40,10 +42,66 @@ export function shouldBlockPipeline(icalData: string | Buffer, now = new Date(),
...
+  };
+}
+
 function containingEventsWithMargin(events: Events, date: Date, advanceMarginSec: number): CalendarEvent[] {
   const bufferedDate = new Date(date.getTime() + advanceMarginSec * 1_000);
+
   return Object.values(events)
     .filter(e => e.type === 'VEVENT')
+    .reduce((arr, e) => {
+      if (e.rrule != null) {
```
Comment (nija-at) [inline change-control-lambda/lib/time-window.ts:96]:
```suggestion
      if (e.rrule) {
```

File: change-control-lambda/lib/time-window.ts [hunk 3]
```diff
@@ -40,10 +42,66 @@ export function shouldBlockPipeline(icalData: string | Buffer, now = new Date(),
...
+ * the event for that date and the following event will be returend
+ * 
+ * If the provided event is not recurring, no events are returned.
+ *
+ * @param event a recurring calendar event.
+ * @param date the date for which the previous and next event should be returned.
+ */
+function getRecurringEvents(event: CalendarEvent, date: Date): CalendarEvent[] {
+  const recurrences: CalendarEvent[] = [];
+  if (event.rrule == null) return recurrences;
```
Comment (nija-at) [inline change-control-lambda/lib/time-window.ts:57]:
```suggestion
  if (!event.rrule) return recurrences;
```

File: change-control-lambda/lib/time-window.ts [hunk 4]
```diff
@@ -40,10 +42,66 @@ export function shouldBlockPipeline(icalData: string | Buffer, now = new Date(),
...
+  const duration = (new Date(event.end).getTime() - new Date(event.start).getTime())
+
+  const after = event.rrule.after(date, false);
```
Comment (nija-at) [inline change-control-lambda/lib/time-window.ts:60]: This tells me that the type of `rrule` is not `any`. What's the correct type that should be assigned here?

File: change-control-lambda/test/time-window.test.ts [hunk 5]
```diff
@@ -30,6 +30,15 @@ DTEND:20190204T075900Z
...
 END:VEVENT
 
+BEGIN:VEVENT
+RRULE:FREQ=WEEKLY;INTERVAL=1
+DTEND:20200504T170000Z
+SUMMARY:Block4
+DTSTAMP:20200501T163641Z
+DTSTART:20200501T220000Z
+SEQUENCE:0
+END:VEVENT
```
Comment (nija-at) [inline change-control-lambda/test/time-window.test.ts:40]: Would prefer if you could create a separate ics input data for these tests so that it's easier to reason when reading and reasoning about the test post-hoc.

Along with that also add comments what this recurrence in plain english (I'm not familiar with the ics format).

File: change-control-lambda/test/time-window.test.ts [hunk 6]
```diff
@@ -58,3 +67,28 @@ test('a blocked window starts AND finishes within margin', () => {
   const x = shouldBlockPipeline(ics, new Date('2019-02-03T07:00:00.000Z'), 72 * 3_600);
   expect(x && x.summary).toBe('Block3');
 });
+
+test('left edge for recurring event after the initial event', () => {
```
Comment (nija-at) [inline change-control-lambda/test/time-window.test.ts:71]: I'm not quite following these test descriptor phrases, and what test cases are being covered.

Can you better phrase, or add any additional comments? What's the initial event and the recurring event here? 
Phrases such as 'current time aligns with the left edge of the first occurrence' might be clearer to understand.

File: change-control-lambda/lib/time-window.ts [hunk 7]
```diff
@@ -40,10 +42,66 @@ export function shouldBlockPipeline(icalData: string | Buffer, now = new Date(),
...
   return Object.values(events)
     .filter(e => e.type === 'VEVENT')
+    .reduce((arr, e) => {
+      if (e.rrule != null) {
+        // Turn a recurrence rule into the events starting on or before
+        // the date, and the next event starting after the date.
+        arr.push(...getRecurringEvents(e, date));
+      } else {
+        arr.push(e);
+      }
```
Comment (nija-at) [inline change-control-lambda/lib/time-window.ts:102]: Would it make sense to push this logic up into `getRecurringEvents()` where it returns the single event in the case it's not recurring. The method can then simply be named `flattenEvents()`

Comment (pnwpedro) [reply change-control-lambda/lib/time-window.ts:102]: I think this suggestion makes sense.

File: change-control-lambda/test/time-window.test.ts (hunk 6 above)
Comment (pnwpedro) [reply change-control-lambda/test/time-window.test.ts:71]: Cool, makes sense. I'll improve the descriptions/add comments.

File: change-control-lambda/test/time-window.test.ts (hunk 5 above)
Comment (pnwpedro) [reply change-control-lambda/test/time-window.test.ts:40]: Will do!

File: change-control-lambda/lib/time-window.ts (hunk 4 above)
Comment (pnwpedro) [reply change-control-lambda/lib/time-window.ts:60]: I'll pull in the type from RRule and add it here.

File: change-control-lambda/lib/time-window.ts (hunk 1 above)
Comment (pnwpedro) [reply change-control-lambda/lib/time-window.ts:55]: I'll change this to flattenEvents() as suggested on the other comment.
